### PR TITLE
R2 solver

### PR DESCRIPTION
**Pull request by nathanemac**
This is a PR for R2 algorithm

### CODE REVIEW

**Comment by dpo** [comment]
Please remove all files that are not part of this repo from your PR.

**Comment by dpo** [comment]
@nathanemac

File: src/R2.jl [hunk 1]
```diff
@@ -0,0 +1,164 @@
+export R2, R2Solver
+
+"""
+    R2(nlp; kwargs...)
+    solver = R2Solver(nlp;)
+    solve!(solver, nlp; kwargs...)
+    A first-order quadratic regularization method for unconstrained optimization
```
Comment (dpo) [inline src/R2.jl:7]:
```suggestion

A first-order quadratic regularization method for unconstrained optimization
```

File: src/R2.jl [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+# Arguments
+- `nlp::AbstractNLPModel{T, V}` is the model to solve, see `NLPModels.jl`.
+
+# Keyword arguments 
+- x::V = nlp.meta.x0`: the initial guess
+- atol = eps(T)^(1 / 3): absolute tolerance
+- rtol = eps(T)^(1 / 3): relative tolerance: algorithm stop when ||∇f(x)|| ≤ ϵ_abs + ϵ_rel*||∇f(x0)||
+- η1 = T(0.3), η2 = T(0.7): step acceptance parameters
```
Comment (dpo) [inline src/R2.jl:16]: The default values in the solver are different.

File: src/R2.jl [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    max_time::Float64 = 3600.0,
+    max_eval::Int = -1,
+    verbose::Bool = true,
+  ) where {T, V}
+
+  unconstrained(nlp) || error("R2 should only be called on unconstrained problems.")
+  start_time = time()
+  elapsed_time = 0.0
+
+  x = solver.x .= copy(x0)
```
Comment (dpo) [inline src/R2.jl:84]:
```suggestion
  x = solver.x .= x0
```

File: src/R2.jl [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  ∇fk = solver.gx
+  ck = solver.cx
+
+  iter = 0
+  fk = obj(nlp, x)
+  
+  grad!(nlp, x, ∇fk)
+  norm_∇fk=norm(∇fk)
```
Comment (dpo) [inline src/R2.jl:92]:
```suggestion
  norm_∇fk = norm(∇fk)
```

File: src/R2.jl [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+  σk = 2^round(log2(norm_∇fk + 1)) # The closest exact-computed power of 2 from ∇fk
```
Comment (dpo) [inline src/R2.jl:93]: Why is this a good initial value?

File: src/R2.jl [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+  # Stopping criterion: 
+  ϵ = atol + rtol*norm_∇fk
```
Comment (dpo) [inline src/R2.jl:96]:
```suggestion
  ϵ = atol + rtol * norm_∇fk
```

File: src/R2.jl [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  if verbose
+    @info @sprintf "%5s  %9s  %7s  %7s " "iter" "f" "‖∇f‖" "σ"
+    infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e" iter fk norm_∇fk σk
+  end
+
+  status = :unknown
+    
+  while !(optimal | tired)
+
+    ck .= x .- (∇fk./σk)
```
Comment (dpo) [inline src/R2.jl:108]:
```suggestion
    ck .= x .- ∇fk ./ σk
```

File: src/R2.jl [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    ΔTk= norm_∇fk^2/ σk
```
Comment (dpo) [inline src/R2.jl:109]:
```suggestion
    ΔTk= norm_∇fk^2 / σk
```

File: src/R2.jl [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  optimal = norm_∇fk ≤ ϵ
+  tired = neval_obj(nlp) > max_eval ≥ 0 || elapsed_time > max_time
+  if verbose
+    @info @sprintf "%5s  %9s  %7s  %7s " "iter" "f" "‖∇f‖" "σ"
+    infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e" iter fk norm_∇fk σk
+  end
+
+  status = :unknown
+    
+  while !(optimal | tired)
```
Comment (dpo) [inline src/R2.jl:106]: If the initial point is optimal, the code doesn't print anything (just the header).

File: test/test_solvers.jl [hunk 10]
```diff
@@ -7,6 +7,7 @@ function tests()
         ("trunk+cg", (nlp; kwargs...) -> trunk(nlp, subsolver_type = CgSolver; kwargs...)),
         ("lbfgs", lbfgs),
         ("tron", tron),
+        ("R2",R2)
```
Comment (dpo) [inline test/test_solvers.jl:10]:
```suggestion
        ("R2", R2)
```

**Review by dpo — CHANGES_REQUESTED** [review]

**Review by dpo — CHANGES_REQUESTED** [review]
@nathanemac there are many polluting files in here.

**Comment by tmigot** [comment]
Hey @nathanemac ! Actually, the tests failed.

In particular, in the `jldoctest` could you modify the examples so that no logs are printed?
Then, the logs to the other failed tests are available here: https://github.com/JuliaSmoothOptimizers/JSOSolvers.jl/runs/7697880907?check_suite_focus=true

**Comment by paraynaud** [comment]
Hey guys, I was wondering if the algorithm is stochastic when it uses a `KnetNLPModel`?
If not you could add a line that would change the current mini-batch for the next `obj, grad` evaluations.
```julia
isa(nlp) == reset_minibatch_train!(nlp)
```
similar to PQNNLPModel in `trunk`:
```julia
if isa(nlp, QuasiNewtonModel)
  ∇fn .-= ∇f
  @. ∇fn = -∇fn  # = ∇f(xₖ₊₁) - ∇f(xₖ)
  push!(nlp, s, ∇fn)
  ∇fn .= ∇f
end
```

The drawback of this is that `KnetNLPModels` must be in `JSOSolvers.jl` dependencies.

**Comment by dpo** [comment]
It should be enough to implement the callback feature like here: https://github.com/JuliaSmoothOptimizers/JSOSolvers.jl/pull/113/files#diff-36c852523b6a7513a10b8a6bae21f1037eaa27484efc9042f75bbda46d9f529bR34

The plan is for all solvers to have a callback. Then a user using R2 and KnetNLPModels can simply call `reset_minibatch_train!()` in the callback.

**Comment by dpo** [comment]
@nathanemac It also looks like all the tests fail here.
@tmigot Could you please help him?

**Comment by nathanemac** [comment]
@dpo ,  problem on tests solved with @tmigot

**Comment by tmigot** [comment]
Great job @nathanemac ! I suggest we add the callback in a separate PR.

**Review by dpo — APPROVED** [review]

**Comment by dpo** [comment]
Many thanks @nathanemac and @tmigot !!!

**Pull request merged**